optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java

```diff
@@ -0,0 +1,219 @@
+import org.optaplanner.core.api.score.stream.ConstraintFactory;
+import org.optaplanner.core.api.score.stream.ConstraintProvider;
+import org.optaplanner.core.api.score.stream.common.ConstraintCollectors;
+import org.optaplanner.core.api.score.stream.common.Joiners;
+import org.optaplanner.examples.machinereassignment.domain.MrMachineCapacity;
+import org.optaplanner.examples.machinereassignment.domain.MrProcessAssignment;
+import org.optaplanner.examples.machinereassignment.domain.solver.MrServiceDependency;
+
+import static org.optaplanner.core.api.score.stream.common.ConstraintCollectors.sumLong;
+
+public class MachineReassignmentConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public void defineConstraints(ConstraintFactory constraintFactory) {
+        maximumCapacity(constraintFactory);
+        serviceConflict(constraintFactory);
+        serviceLocationSpread(constraintFactory);
+        serviceDependency(constraintFactory);
+        transientUsage(constraintFactory);
+
+        loadCost(constraintFactory);
+        balanceCost(constraintFactory);
+        processMoveCost(constraintFactory);
+        serviceMoveCost(constraintFactory);
+        machineMoveCost(constraintFactory);
+    }
+
+    /*************** HARD ***************/
+
+    /**
+     * Maximum capacity: The maximum capacity for each resource for each machine must not be exceeded.
+     */
+    private void maximumCapacity(ConstraintFactory constraintFactory) {
+        Constraint constraint = constraintFactory.newConstraintWithWeight(MrConstraintName.MAXIMUM_CAPACITY.getName(), HardSoftLongScore.ofHard(1L));
+        constraint.from(MrMachineCapacity.class)
+                .join(MrProcessAssignment.class,
+                      Joiners.equalTo(MrMachineCapacity::getMachine, MrProcessAssignment::getMachine)
+                )
+                .groupBy(
+                        (machineCapacity, processAssignment) -> machineCapacity, sumLong(
+                                (machineCapacity, processAssignment) -> processAssignment.getUsage(machineCapacity.getResource())
+                        )
+                )
+                .filter(((machineCapacity, usage) -> machineCapacity.getMaximumCapacity() < usage))
+                .penalizeLong((machineCapacity, usage) -> machineCapacity.getMaximumCapacity() - usage);
+    }
+
+    /**
+     * Conflict: Processes of the same service must run on distinct machines.
+     */
+    private void serviceConflict(ConstraintFactory constraintFactory) {
+        Constraint constraint = constraintFactory.newConstraintWithWeight(MrConstraintName.SERVICE_CONFLICT.getName(),
+                                                                          HardSoftLongScore.ofHard(1L));
+        constraint.from(MrProcessAssignment.class)
+                .joinOther(
+                        Joiners.equalTo(MrProcessAssignment::getMachine),
+                        Joiners.lessThan(MrProcessAssignment::getId)
```

Comment:
joinOther() implies the lessThan(MrProcessAssignment::getId) automatically, see javadoc

Reply:
Ok, that javadoc doesn't really clearly mention that. I improving that now :)

Reply:
javadoc improved on my branch, rebase on to confirm the javadoc is now clearer :)